src/crystal/hasher.cr

```diff
@@ -45,12 +94,53 @@ struct Crystal::Hasher
   end
 
   def reference(value)
-    @result = @result * 31 + value.object_id.to_u64
-    self
+    permute(value.object_id.to_u64)
   end
 
   def string(value)
-    value.to_slice.hash(self)
+    bytes(value.to_slice)
+  end
+
+  private def read_u24(u, r)
+    u[0].to_u64 | (u[r/2].to_u64 << 8) | (u[r - 1].to_u64 << 16)
+  end
+
+  private def read_u32(u)
+    # force correct unaligned read
+    t4 = uninitialized UInt32
+    pointerof(t4).as(UInt8*).copy_from(u, 4)
+    t4.to_u64
+  end
+
+  private def read_u64(u)
+    # force correct unaligned read
+    t8 = uninitialized UInt64
+    pointerof(t8).as(UInt8*).copy_from(u, 8)
+    t8
+  end
+
+  def bytes(value)
+    bsz = value.size
```

Comment:
Why not `byte_size` or just `size`? `bsz` is unnecessarily cryptic.

Reply:
done